src/main/java/org/skriptlang/skript/bukkit/recipes/elements/ExprRecipeIngredients.java

```diff
@@ -0,0 +1,473 @@
+import org.jetbrains.annotations.Nullable;
+import org.skriptlang.skript.bukkit.recipes.MutableRecipe;
+import org.skriptlang.skript.bukkit.recipes.MutableRecipe.*;
+import org.skriptlang.skript.bukkit.recipes.MutableRecipe.MutableCraftingRecipe.MutableShapedRecipe;
+import org.skriptlang.skript.bukkit.recipes.MutableRecipe.MutableSmithingRecipe.MutableSmithingTransformRecipe;
+import org.skriptlang.skript.bukkit.recipes.MutableRecipe.MutableSmithingRecipe.MutableSmithingTrimRecipe;
+import org.skriptlang.skript.bukkit.recipes.CreateRecipeEvent;
+import org.skriptlang.skript.bukkit.recipes.CreateRecipeEvent.*;
+import org.skriptlang.skript.bukkit.recipes.CreateRecipeEvent.CraftingRecipeEvent.ShapedRecipeEvent;
+import org.skriptlang.skript.bukkit.recipes.CreateRecipeEvent.SmithingRecipeEvent.SmithingTransformRecipeEvent;
+import org.skriptlang.skript.bukkit.recipes.CreateRecipeEvent.SmithingRecipeEvent.SmithingTrimRecipeEvent;
+
+import java.util.*;
+
+@Name("Recipe Ingredients")
+@Description({
+	"The ingredients of a shaped or shapeless recipe.",
+	"The ingredients of a row for a shaped recipe.",
+	"The input item of a blasting, furnace, campfire, smoking, stonecutting and transmute recipe.",
+	"The base, template, addition items of a smithing transform and smithing trim recipe.",
+	"the transmute items of a transmute recipe."
+})
+@Examples({
+	"set {_recipe} to a new shaped recipe with the key \"my_recipe\":",
+		"\tset the recipe ingredients to diamond, air, diamond, air, emerald, air, diamond, air and diamond",
+		"\t#OR",
+		"\tset the recipe ingredients of the 1st row to diamond, air and diamond",
+		"\tset the recipe ingredients of second row to air, emerald and air",
+		"\tset the recipe ingredients of the third row to diamond, air and diamond",
+		"\tset the recipe result to beacon",
+	"",
+	"set {_recipe} to a shapeless recipe with id \"my_recipe\":",
+		"\tset recipe ingredients to iron ingot, gold ingot, iron ingot, nether star, 5 obsidian, nether star, iron ingot, gold ingot and iron ingot",
+		"\tset recipe resulting item to beacon named \"OP Beacon\"",
+	"",
+	"set {_recipe} to new blasting recipe with the id \"my_recipe\":",
+		"\tset the recipe input item to netherite ingot named \"Impure Netherite\"",
+		"\tset the recipe result item to netherite ingot named \"Pure Netherite\"",
+	"",
+	"set {_recipe} to a new smithing transform recipe with key \"my_recipe\":",
+		"\tset the recipe base item to diamond helmet",
+		"\tset the recipe template item to paper named \"Blueprint\"",
+		"\tset the recipe addition item to netherite ingot named \"Pure Netherite\"",
+		"\tset the recipe result to netherite helmet named \"Pure Helmet\"",
+	"",
+	"set {_recipe} to a new transmute recipe with key \"my_recipe\":",
+		"\tset the recipe input item to leather helmet",
+		"\tset the recipe transmute item to nether star named \"Free Upgrade\"",
+		"\tset the recipe result to netherite helmet"
```

Comment:
i know it's probably already been discussed but do you think there's any safe way to get rid of all the `recipe` keywords here? It feels awkward to have to repeat them

Reply:
`RecipeCategory` - I could
`RecipeCookingTime` - Can't because of `ExprBurnCookTime`
`RecipeExperience` - Not sure, might conflict with `ExprExperience` and any event values
`RecipeGroup` - Might conflict with `ExprGroup`
`RecipeIngredients`:
```
Ingredients - Changed
Rows - Changed
Input - Might conflict with "ExprInput"
Base - Already optional
Template - Already optional
Addition - Already optional
Transmute - Already optional
```
`RecipeResult` - Might conflict with `ExprFurnaceSlot`
`RecipeKey` - Should be fine

Let me know

Reply:
yeah sounds good